sql/core/src/test/scala/org/apache/spark/sql/TPCDSQueryTestSuite.scala

```diff
@@ -97,50 +97,91 @@ class TPCDSQueryTestSuite extends QueryTest with TPCDSBase with SQLQueryTestHelp
+              "\n"
+          }
+          val parent = goldenFile.getParentFile
+          if (!parent.exists()) {
+            assert(parent.mkdirs(), "Could not create directory: " + parent)
+          }
+          stringToFile(goldenFile, goldenOutput)
+        }
+
+        // Read back the golden file.
+        val (expectedSchema, expectedOutput) = {
+          val goldenOutput = fileToString(goldenFile)
+          val segments = goldenOutput.split("-- !query.*\n")
+
+          // query has 3 segments, plus the header
+          assert(segments.size == 3,
+            s"Expected 3 blocks in result file but got ${segments.size}. " +
+              "Try regenerate the result files.")
+
+          (segments(1).trim, segments(2).replaceAll("\\s+$", ""))
+        }
+
+        assertResult(expectedSchema, s"Schema did not match\n$queryString") {
+          schema
+        }
+        if (needSort) {
+          val expectSorted = expectedOutput.split("\n").sorted.map(_.trim)
+            .mkString("\n").replaceAll("\\s+$", "")
+          val outputSorted = output.sorted.map(_.trim).mkString("\n").replaceAll("\\s+$", "")
+          assertResult(expectSorted, s"Result did not match\n$queryString") {
+            outputSorted
+          }
+        } else {
+          assertResult(expectedOutput, s"Result did not match\n$queryString") {
+            outputString
+          }
+        }
+      } catch {
+        case e: Throwable =>
+          val configs = conf.map {
+            case (k, v) => s"$k=$v"
+          }
+          throw new Exception(s"${e.getMessage} \nError using configs:\n${configs.mkString("\n")}")
       }
-      val parent = goldenFile.getParentFile
-      if (!parent.exists()) {
-        assert(parent.mkdirs(), "Could not create directory: " + parent)
-      }
-      stringToFile(goldenFile, goldenOutput)
     }
+  }
 
-    // Read back the golden file.
-    val (expectedSchema, expectedOutput) = {
-      val goldenOutput = fileToString(goldenFile)
-      val segments = goldenOutput.split("-- !query.*\n")
+  val sortMergeJoinConf = Map(
+    SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1",
+    SQLConf.PREFER_SORTMERGEJOIN.key -> "true")
 
-      // query has 3 segments, plus the header
-      assert(segments.size == 3,
-        s"Expected 3 blocks in result file but got ${segments.size}. " +
-          "Try regenerate the result files.")
+  val broadcastHashJoinConf = Map(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "10485760")
 
-      (segments(1).trim, segments(2).replaceAll("\\s+$", ""))
-    }
+  val shuffleHashJoinConf = Map(
+    SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1",
+    "spark.sql.join.forceApplyShuffledHashJoin" -> "true")
 
-    assertResult(expectedSchema, s"Schema did not match\n$queryString") { schema }
-    assertResult(expectedOutput, s"Result did not match\n$queryString") { outputString }
-  }
+  val joinConfSet: Set[Map[String, String]] =
+    Set(broadcastHashJoinConf, shuffleHashJoinConf);
```

Comment:
nit: it's a bit weird that `joinConfSet` doesn't include `sortMergeJoinConf`. One idea is
```
runQuery(queryString, goldenFile, joinConfSet.head, false)
if (!regenerateGoldenFiles) {
  joinConfSet.tail.foreach { conf =>
    runQuery(queryString, goldenFile, conf.toSeq, true)
  }
}
```

Reply:
OK

Reply:
Thanks.